In this Notebook, I plan to explore how the fantasy genre has been dominated by mostly male authors, and how that is slowly chaning over time. Lets begin!

In [48]:
import pandas as pd
import altair as alt
import matplotlib.pyplot as plt


In [49]:
novels_df = pd.read_csv('https://raw.githubusercontent.com/melaniewalsh/responsible-datasets-in-context/main/datasets/top-500-novels/library_top_500.csv')
nyt_bestsellers_df = pd.read_csv("https://raw.githubusercontent.com/ecds/post45-datasets/main/nyt_full.tsv", sep='\t')
nyt_bestsellers_df = nyt_bestsellers_df.rename(columns={'title': 'nyt_title'})
nyt_bestsellers_df['title'] = nyt_bestsellers_df['nyt_title'].str.capitalize()
combined_novels_nyt_df = novels_df.merge(nyt_bestsellers_df, how='left', on=['author', 'title'])
combined_novels_nyt_df.columns


Index(['top_500_rank', 'title', 'author', 'pub_year', 'orig_lang', 'genre',
       'author_birth', 'author_death', 'author_gender', 'author_primary_lang',
       'author_nationality', 'author_field_of_activity', 'author_occupation',
       'oclc_holdings', 'oclc_eholdings', 'oclc_total_editions',
       'oclc_holdings_rank', 'oclc_editions_rank', 'gr_avg_rating',
       'gr_num_ratings', 'gr_num_reviews', 'gr_avg_rating_rank',
       'gr_num_ratings_rank', 'oclc_owi', 'author_viaf', 'gr_url', 'wiki_url',
       'pg_eng_url', 'pg_orig_url', 'year', 'week', 'rank', 'title_id',
       'nyt_title'],
      dtype='object')

After combining the datasets of the Top 500 Greatest Novels and the New York Times Best Sellers list, we have a list of 721 books and authors that were in both sets of the data. 

In [50]:
print(combined_novels_nyt_df['author'].nunique())

279


In [51]:
combined_novels_nyt_df['pub_year'] = combined_novels_nyt_df['pub_year'].astype(int)
combined_novels_nyt_df['author_gender'] = combined_novels_nyt_df['author_gender'].astype(str)

In [52]:
combined_novels_nyt_df = combined_novels_nyt_df.dropna(subset=['pub_year'])

valid_years = combined_novels_nyt_df[(combined_novels_nyt_df['pub_year'] >= 1800) & (combined_novels_nyt_df['pub_year'] <= pd.Timestamp.now().year)]

# Aggregate the data
# Group by year and gender, and count the number of authors
author_counts = valid_years.groupby(['pub_year', 'author_gender']).size().reset_index(name='count')

author_counts['pub_year'] = pd.to_datetime(author_counts['pub_year'], format='%Y')
# Create the Altair chart
chart = alt.Chart(author_counts).mark_line().encode(
    x=alt.X('pub_year:T', title='Year'),
    y=alt.Y('count:Q', title='Number of Authors'),
    color=alt.Color('author_gender:N', title='Gender')
).properties(
    title='Gender Disparity in Authors Over Time'
)

# Display the chart
chart

alt.Chart(...)

As you can see, this graph ended up being a little wonkey. Lets retry with a scatterplot. 

In [53]:
scatter_plot = alt.Chart(author_counts).mark_circle(size=60).encode(
    x=alt.X('pub_year:T', title='Year'),
    y=alt.Y('count:Q', title='Number of Authors'),
    color=alt.Color('author_gender:N', title='Gender'),
    tooltip=['pub_year:T', 'count:Q', 'author_gender:N']
).properties(
    title='Gender Disparity in Authors Over Time'
)

# Display the scatter plot
scatter_plot

alt.Chart(...)

This scatterplot still doesn't tell us too much about the gender of popular authors. We can see some clear outlier years more recently, and with a couple exeptions like 1938 having a lot of female authors. Time to look a little deeper into 1938 to see what caused it to have so many female authors. 

In [54]:
female_authors_1938 = combined_novels_nyt_df[(combined_novels_nyt_df['author_gender'] == 'female') & (combined_novels_nyt_df['pub_year'] == 1938)]
print(female_authors_1938)

     top_500_rank         title                    author  pub_year orig_lang  \
167           130       Rebecca         Daphne Du Maurier      1938   English   
168           130       Rebecca         Daphne Du Maurier      1938   English   
169           130       Rebecca         Daphne Du Maurier      1938   English   
170           130       Rebecca         Daphne Du Maurier      1938   English   
171           130       Rebecca         Daphne Du Maurier      1938   English   
172           130       Rebecca         Daphne Du Maurier      1938   English   
173           130       Rebecca         Daphne Du Maurier      1938   English   
174           130       Rebecca         Daphne Du Maurier      1938   English   
175           130       Rebecca         Daphne Du Maurier      1938   English   
176           130       Rebecca         Daphne Du Maurier      1938   English   
177           130       Rebecca         Daphne Du Maurier      1938   English   
178           130       Rebe

It seems this book was listed many times in this dataset. Although interesting to note, it won't help us with looking at the gender of authors over time. More data cleaning must be done.

In [66]:
unique_authors_df = combined_novels_nyt_df.drop_duplicates(subset=['author'])
unique_authors_df

top_500_rank                               title                  author  \
0               1                         Don Quixote     Miguel de Cervantes   
1               2    Alice's Adventures in Wonderland           Lewis Carroll   
2               3  The Adventures of Huckleberry Finn              Mark Twain   
4               5                     Treasure Island  Robert Louis Stevenson   
5               6                 Pride and Prejudice             Jane Austen   
..            ...                                 ...                     ...   
697           494              The Naked and the Dead           Norman Mailer   
699           496          Stranger in a Strange Land      Robert A. Heinlein   
700           497                     Vision in White            Nora Roberts   
701           498                    The Whipping Boy          Sid Fleischman   
702           499                                Room           Emma Donoghue   

     pub_year orig_lang    genre author_birth author_death author_gender  \
0        1605   Spanish   action         1547         1616          male   
1        1865   English  fantasy         1832         1898          male   
2        1884   English   action         1835         1910          male   
4        1883   English   action         1850         1894          male   
5        1813   English  romance         1775         1817        female   
..        ...       ...      ...          ...          ...           ...   
697      1948   English  history         1923         2007          male   
699      1961   English    scifi         1907         1988          male   
700      2009   English  romance         1965        ALIVE        female   
701      1986   English   action         1920         2010          male   
702      2010   English       na         1969        ALIVE        female   

    author_primary_lang  ... author_viaf  \
0                   spa  ...    17220427   
1                   eng  ...    66462036   
2                   eng  ...    50566653   
4                   eng  ...    95207986   
5                   eng  ...   102333412   
..                  ...  ...         ...   
697                 eng  ...     7393743   
699                 eng  ...    12309757   
700                 eng  ...    66448023   
701                 eng  ...    66438084   
702                 eng  ...    39539889   

                                                gr_url  \
0    https://www.goodreads.com/book/show/3836.Don_Q...   
1    https://www.goodreads.com/book/show/24213.Alic...   
2    https://www.goodreads.com/book/show/2956.The_A...   
4    https://www.goodreads.com/book/show/295.Treasu...   
5    https://www.goodreads.com/book/show/1885.Pride...   
..                                                 ...   
697                                                NaN   
699                                                NaN   
700                                                NaN   
701                                                NaN   
702                                                NaN   

                                              wiki_url  \
0            https://en.wikipedia.org/wiki/Don_Quixote   
1    https://en.wikipedia.org/wiki/Alice%27s_Advent...   
2    https://en.wikipedia.org/wiki/Adventures_of_Hu...   
4        https://en.wikipedia.org/wiki/Treasure_Island   
5    https://en.wikipedia.org/wiki/Pride_and_Prejudice   
..                                                 ...   
697  https://en.wikipedia.org/wiki/The_Naked_and_th...   
699  https://en.wikipedia.org/wiki/Stranger_in_a_St...   
700      https://en.wikipedia.org/wiki/Vision_in_White   
701     https://en.wikipedia.org/wiki/The_Whipping_Boy   
702         https://en.wikipedia.org/wiki/Room_(novel)   

                                            pg_eng_url  \
0    https://www.gutenberg.org/cache/epub/996/pg996...   
1     https://www.gutenberg.org/cache/epub/11/pg11.txt   
2     https://www.gute

We now have a dataset with all unique authors in the datasets, wiht a total of 279. Although looking at this data is not as complete as looking at every book by every author, it will give us a cleaner graph to look at author's genders based on when they wrote the book they were selected for. Lets make another scatterplot to compare and see how different it looks. 

In [67]:
unique_authors_df['pub_year'] = unique_authors_df['pub_year'].astype(str)
unique_authors_df['pub_year'] = pd.to_datetime(unique_authors_df['pub_year'], format='%Y', errors='coerce')
scatter_plot = alt.Chart(unique_authors_df).mark_circle(size=60).encode(
    x=alt.X('pub_year:T', title='Year'),
    y=alt.Y('count()', title='Number of Authors'),
    color=alt.Color('author_gender:N', title='Gender'),
    tooltip=['pub_year:T', 'count()', 'author_gender:N']
).properties(
    title='Gender Disparity in Authors Over Time'
)

# Display the scatter plot
scatter_plot

/var/folders/fp/fz7jly1s1nd0k1gc00ph9l680000gn/T/ipykernel_55692/260561285.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  unique_authors_df['pub_year'] = unique_authors_df['pub_year'].astype(str)
/var/folders/fp/fz7jly1s1nd0k1gc00ph9l680000gn/T/ipykernel_55692/260561285.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  unique_authors_df['pub_year'] = pd.to_datetime(unique_authors_df['pub_year'], format='%Y', errors='coerce')


alt.Chart(...)

This scatterplot is a little more descriptive. We can see there is some years were 2 female authors made the dataset, but only 1 year (2009) 3 female authors made it, and none for 4 or 5. Now we can start to see the gender disparity between authors in every genre, it's time to focus in on the fantasy genre. 

In [68]:
fantasy_unique_authors_df = unique_authors_df[unique_authors_df['genre'] == 'fantasy']
fantasy_unique_authors_df

top_500_rank                                  title               author  \
1               2       Alice's Adventures in Wonderland        Lewis Carroll   
10             11                     Gulliver's Travels       Jonathan Swift   
17             18   The Hobbit, or, There and Back Again       J.R.R. Tolkien   
29             30                       The Wizard of Oz        L. Frank Baum   
38             39                The Wind in the Willows      Kenneth Grahame   
44             45  Harry Potter and the Sorcerer's Stone         J.K. Rowling   
58             51  The Lion, the Witch, and the Wardrobe           C.S. Lewis   
66             59                              Peter Pan          J.M. Barrie   
104            97                        Charlotte's Web           E.B. White   
251           183                   Bridge to Terabithia   Katherine Paterson   
301           233                       Tuck Everlasting      Natalie Babbitt   
342           274                                 Eragon  Christopher Paolini   
406           325                    The Lightning Thief         Rick Riordan   
426           345                           Mary Poppins         P.L. Travers   
427           346                           Artemis Fowl          Eoin Colfer   
431           350                     The Dark Is Rising         Susan Cooper   
445           364                      A Game of Thrones   George R.R. Martin   
524           381             The Indian in the Cupboard     Lynne Reid Banks   
558           415       Mrs. Frisby and the Rats of NIMH    Robert C. O'Brien   
591           419               The Once and Future King          T. H. White   
616           444                           Lost Horizon         James Hilton   
620           448                               Inkheart       Cornelia Funke   
658           472                  The Phantom Tollbooth        Norton Juster   

      pub_year orig_lang    genre author_birth author_death author_gender  \
1   1865-01-01   English  fantasy         1832         1898          male   
10  1726-01-01   English  fantasy         1667         1745          male   
17  1937-01-01   English  fantasy         1892         1973          male   
29  1900-01-01   English  fantasy         1856         1919          male   
38  1908-01-01   English  fantasy         1859         1932          male   
44  1997-01-01   English  fantasy         1965        ALIVE        female   
58  1950-01-01   English  fantasy         1898         1963          male   
66  1911-01-01   English  fantasy         1860         1937          male   
104 1952-01-01   English  fantasy         1899         1985          male   
251 1977-01-01   English  fantasy         1932        ALIVE        female   
301 1975-01-01   English  fantasy         1932         2016        female   
342 2002-01-01   English  fantasy         1983        ALIVE          male   
406 2005-01-01   English  fantasy         1964        ALIVE          male   
426 1934-01-01   English  fantasy         1899         1996        female   
427 2001-01-01   English  fantasy         1965        ALIVE          male   
431 1973-01-01   English  fantasy         1935        ALIVE        female   
445 1996-01-01   English  fantasy         1948        ALIVE          male   
524 1980-01-01   English  fantasy         1929        ALIVE        female   
558 1971-01-01   English  fantasy         1918         1973          male   
591 1958-01-01   English  fantasy         1906         1964           nan   
616 1933-01-01   English  fantasy         1900         1954          male   
620 2003-01-01    German  fantasy         1958        ALIVE        female   
658 1961-01-01   English  fantasy         1929         2021          male   

    author_primary_lang  ... author_viaf  \
1                   eng  ...    66462036   
10                  eng  ...    14777110   
17                  eng  ...    95218067   
29                  eng  ...     49

This narrows us down to 23 fantasy authors, with mostly books you probably recognize. You can eyeball the data and see it's mostly male dominated, but we can make one more graph to help us vizualize it. 

In [69]:
gender_counts = fantasy_unique_authors_df['author_gender'].value_counts().reset_index()
gender_counts.columns = ['author_gender', 'count']

# Create the Altair bar chart
bar_chart = alt.Chart(gender_counts).mark_bar().encode(
    x=alt.X('author_gender:N', title='Gender'),
    y=alt.Y('count:Q', title='Number of Authors'),
    color=alt.Color('author_gender:N', title='Gender'),
    tooltip=['author_gender:N', 'count:Q']
).properties(
    title='Number of Unique Authors by Gender'
)

# Display the bar chart
bar_chart

alt.Chart(...)


The bar chart shows 16 males to 7 females, as after further inspection the one NAN is TH White, a male english author. Although not an overwhelming majority, it is clear fantasy is a mostly male dominated field for authors. Especially in the past, as most of the woman on the fantasy list have released their book within the last 50 years, with the exeception of The Indian in the Cupboard by Lynne Reid Banks. 